### PR TITLE
Explosions and Voltage: A Complete Powernet Overhaul

### PR DESCRIPTION
**Pull request by Sirryan2002**
## What Does This PR Do
Completely overhauls the powernet by giving it the largest update it will have (not lying here) in over a decade.

**Notable Changes**

- Power is no longer standard over every machine and engine, now there are two types of power: **Low Voltage** and **High Voltage**. Both require their own new set of cables respectively named Low Voltage Cables and High Voltage Cables
- Most engines will produce High Voltage power which will need to be distributed throughout the station and converted to Low Voltage power via transformers so that machines in regional low-voltage powernets can be powered.
- MAPPING: Each region of the station will be broken up into its own regional powernet and will each come with a power maintenance room with a battery, grid monitoring computer, transformer, and machine breaker box
- These new breaker boxes contain the breakers for every APC on the net, as well as a master switch for the entire regional powernet. These will be great for antagonists trying to sabotage certain areas or break in with a crowbar. The master switch is locked in place via a lock which you'll need special tools to get passed.
- Explosions, APC overloading, and wire melting when you accidentally mix high voltage and low voltage power together

## Why It's Good For The Game
see #19786

Power is currently hell to deal with and not at all balanced. It's very much not beginner friendly and lacks a lot of logical middle steps when trying to get engines to power an entire station. A few important things this change will do:

- This will add much needed content to power mechanics for engineering
- make problem-solving the powernet more fun and diagnosable while giving antags the ability to more specifically target certain parts of the station for sabotage.
- Breaks up the global powernet into manageable pieces which will be easier to work with as an engineer and allow engineers to more easily modularize their powernets
- sets the foundations for much more impactful changes down the road #19734

## WIP PR
There's still much to be done. A lot of the under-the-hood code for LV-HV power still needs to be done (Fuck powercode tbh) and I still need to complete the mapping changes.

- [x] Refactor Regional Powernets to be able to handle different voltage types
- [x] Refactor powernet propagation to creation separate HV & LV nets
   - [ ] Load Limiters
- [x] Make engines use HV power 
- [x] Make powernets automatically pickup voltage incompatibilities and track them with datums
- [ ] Make Transformers Work
- [x] Make HV Connectors properly link machines to powernets
- [ ] Make Breakers Work
- [x] Breaker Box TGUI (it's gonna look fucking cool)
- [x] Make Batteries work
   - [ ] New Accumulators
   - [ ] SMES units
   - [x] Hook into the process power cycle
- [ ] Specialized HV Construction
   - [x] Boltcutters
   - [ ] New Construction Paths + Components
- [ ] Power Overloads
   - [ ] APC Arcing rework
   - [ ] Electricity sparks and wire melting effects
   - [ ] Increased damage/effects for messing with HV stuff
- [ ] Integrate stuff in Power Monitoring Consoles
- [ ] Station Mapping (help me god)
- [ ] Spriting stuff
   - [X] Transformers
   - [x] Transformer Animated state?
   - [X] Accumulators
   - [ ] Accumulator overlays & smooth
   - [X] Breaker Box Closed Stat
   - [ ] Breaker Box Open State
   - [X] HV-Cables
   - [ ] Load Limiter
- [ ] Add ion_act, obj_break, and other acts to each new machines
- [ ] Add admin logging & buttons for each machine

## Changelog
:cl:
experiment: fucked around and found out
/:cl:

### CODE REVIEW

**Comment by ArcticValor** [comment]
oh no, more engi stuff I have to learn? (I'm already struggling to comprehend atmos)

**Comment by Venuska1117** [comment]
do cutting high voltage cable without any safty measures will make you get shocked on similar level like defibrillator? considering we split them now to low and high voltage ones?

**Comment by Sirryan2002** [comment]
yes, except it will be much more violent

File: code/modules/power/hv_connector.dm [hunk 1]
```diff
@@ -0,0 +1,58 @@
+
+/obj/machinery/power/hv_connector
+	name = "high-voltage power connector"
+	desc = "a large plug which handles the transfer of high-voltage power from engines into cabling"
```
Comment (lewcc) [inline code/modules/power/hv_connector.dm:4]:
```suggestion
	desc = "A large plug which handles the transfer of high-voltage power from engines into cabling"
```

File: code/modules/power/power_machine.dm [hunk 2]
```diff
@@ -1,34 +1,65 @@
...
+	/// The voltage type this machine can properly interface with on powernets
+	var/power_voltage_type = VOLTAGE_LOW
+	/// The default method that this power machine will use to connect to the powernet
+	var/powernet_connection_type = PW_CONNECTION_NODE
+
+	var/list/linked_connectors = list()
+	/// The regional powernet this power machine is hooked into
 	var/datum/regional_powernet/powernet = null
 
+	///
```
Comment (lewcc) [inline code/modules/power/power_machine.dm:25]: incomplete comment

File: code/modules/power/power_machine.dm [hunk 3]
```diff
@@ -87,23 +139,16 @@
...
-			return C
-	return null
-
-/area/proc/get_apc()
-	for(var/thing in GLOB.apcs)
-		var/obj/machinery/power/apc/APC = thing
-		if(APC.apc_area == src)
-			return APC
+/// Electricity fucking hurts! Let's make sure players understand that
+/obj/machinery/power/proc/attack_zap_check(mob/living/user)
```
Comment (lewcc) [inline code/modules/power/power_machine.dm:151]: Why bother checking the type if you're gonna return the same thing?

File: code/modules/power/powernets/regional_powernet.dm [hunk 4]
```diff
@@ -91,6 +105,19 @@
...
+/datum/regional_powernet/proc/update_voltage(new_voltage)
+	if(!new_voltage)
+		stack_trace("update_voltage called with a null new_voltage type")
+	if(!isnull(power_voltage_type) && new_voltage != power_voltage_type)
+		bzzzt(new_voltage)
+		return FALSE
+	power_voltage_type = new_voltage
+	return TRUE
+
+/datum/regional_powernet/proc/bzzzt(new_voltage)
```
Comment (lewcc) [inline code/modules/power/powernets/regional_powernet.dm:119]: What does this do?

**Review by lewcc — COMMENTED** [review]
First pass

File: code/modules/power/powernets/regional_powernet.dm (hunk 4 above)
Comment (Sirryan2002) [reply code/modules/power/powernets/regional_powernet.dm:119]: was gonna be something with short circuits, decided to do something different and accidentally left this in. Remove in next commit.

File: code/modules/power/power_machine.dm (hunk 3 above)
Comment (Sirryan2002) [reply code/modules/power/power_machine.dm:151]: need this to return these values while in testing, I still haven't fully implemented zapping yet

File: code/modules/power/power_machine.dm (hunk 2 above)
Comment (Sirryan2002) [reply code/modules/power/power_machine.dm:25]: fixed in next commit

File: code/modules/power/cables/cable_coil.dm [hunk 5]
```diff
@@ -29,42 +24,15 @@ GLOBAL_LIST_INIT(cable_coil_recipes, list (new/datum/stack_recipe/cable_restrain
 	usesound = 'sound/items/deconstruct.ogg'
 	toolspeed = 1
 
-/obj/item/stack/cable_coil/New(location, length = MAXCOIL, paramcolor = null)
-	. = ..()
-	if(paramcolor)
-		color = paramcolor
+	var/cable_structure_type = null
```
[the remark this answers is not in the document]
Comment (Sirryan2002) [reply code/modules/power/cables/cable_coil.dm:27]: done in next commit

File: code/modules/power/batteries/battery.dm [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+/obj/machinery/power/battery
+	name = "power battery"
+	desc = "A power container which can save power over long periods of time"
+	icon_state = "smes"
+	density = TRUE
+
+
+	powernet_connection_type = PW_CONNECTION_NODE
+	//
```
[the remark this answers is not in the document]
Comment (Sirryan2002) [reply code/modules/power/batteries/battery.dm:16]: resolved in next commit

File: code/modules/power/batteries/battery.dm [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+	var/charge = 0 // actual charge
+	var/max_capacity = DEFAULT_SAFE_CAPACITY * 2
+	var/battery_status = "FUCK"
```
[the remark this answers is not in the document]
Comment (Sirryan2002) [reply code/modules/power/batteries/battery.dm:19]: cleaned up in next commit

File: code/modules/power/batteries/battery.dm [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+	powernet.batteries |= src
+
+/obj/machinery/power/battery/disconnect_from_network()
+	powernet.batteries -= src
+	return ..()
+
+
+/// will remove charge from the battery equal to the amount specified or up to what charge is left in the battery, returns amount of energy consumed
+/obj/machinery/power/battery/proc/consume_charge(amount)
+	var/amount_consumed =  clamp(round(amount), 0, charge) // can't consume less than 0, can't consume more than the current charge
```
[the remark this answers is not in the document]
Comment (Sirryan2002) [reply code/modules/power/batteries/battery.dm:38]: fixed in next commit

File: code/modules/power/breaker_box.dm [hunk 9]
```diff
@@ -0,0 +1,183 @@
+#define MASTER_BREAKER_
```
[the remark this answers is not in the document]
Comment (Sirryan2002) [reply code/modules/power/breaker_box.dm:1]: correct, I'm still working on breaker boxes

File: code/modules/power/breaker_box.dm [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+
+#define BREAKER_DELAY (1 SECONDS)
```
[the remark this answers is not in the document]
Comment (Sirryan2002) [reply code/modules/power/breaker_box.dm:3]: fixed in next commit

File: code/modules/mining/ores_coins.dm [hunk 11]
```diff
@@ -450,7 +450,7 @@ GLOBAL_LIST_INIT(sand_recipes, list(\
 			..()
 			return
 
-		var/obj/item/stack/cable_coil/CC = new/obj/item/stack/cable_coil(user.loc)
+		var/obj/item/stack/cable_coil/low_voltage/CC = new/obj/item/stack/cable_coil/low_voltage(user.loc)
```
[the remark this answers is not in the document]
Comment (Sirryan2002) [reply code/modules/mining/ores_coins.dm:453]: fixed in next commit

File: code/game/objects/items/tools/boltcutter.dm [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+	hitsound = 'sound/items/wirecutter.ogg'
+	usesound = 'sound/items/wirecutter.ogg'
+	drop_sound = 'sound/items/handling/wirecutter_drop.ogg'
+	pickup_sound = 'sound/items/handling/wirecutter_pickup.ogg'
+	sharp = TRUE
+	toolspeed = 1
+	tool_behaviour = TOOL_WIRECUTTER
+
+
+/obj/item/wirecutters/New(loc, param_color = null)
```
[the remark this answers is not in the document]
Comment (Sirryan2002) [reply code/game/objects/items/tools/boltcutter.dm:24]: fixed

**Comment by matttheficus** [comment]
SoP Changes Needed:
- Chief Engineer
- Station Engineer

Likely needs SoP regarding safe operating parameters for high/low voltage and such. Might be worth re-looking at a large amount of Engi SoP with this big of a change.

**Comment by Sirryan2002** [comment]
This will be reopened in the future, I'm just busy with school.

**Pull request closed without merging**